game-core/src/main/java/games/strategy/engine/data/GameMap.java

```diff
@@ -127,19 +137,41 @@ public Territory getTerritory(final String s) {
   /**
    * Returns all neighbors within a certain distance of the starting territory that match the
    * condition. Does NOT include the original/starting territory in the returned Set.
+   *
+   * @param territory starting territory
+   * @param distance distance to check
+   * @param cond the neighbor territory must match this condition
    */
   public Set<Territory> getNeighbors(
-      final Territory territory, final int distance, final Predicate<Territory> cond) {
+      final Territory territory, final int distance, @Nullable final Predicate<Territory> cond) {
+    return getNeighbors(territory, distance, cond, (it, it2) -> true);
+  }
+
+  /**
+   * Returns all neighbors within a certain distance of the starting territory that match the
+   * condition. Does NOT include the original/starting territory in the returned Set.
+   *
+   * @param territory starting territory
+   * @param distance distance to check
+   * @param cond the neighbor territory must match this condition
+   * @param routeCond the route between the starting territory and the neighbor territory must match
+   *     this condition
+   */
+  public Set<Territory> getNeighbors(
+      final Territory territory,
+      final int distance,
+      @Nullable final Predicate<Territory> cond,
+      final BiPredicate<Territory, Territory> routeCond) {
```

Comment:
Thank you for the previous cleanups, it's now clear that either `cond` is used with an always true bi-predicate (route condition), or if the bi-predicate is used then it is always with these two conditions:
```
                ProMatches.territoryCanPotentiallyMoveLandUnits(player, data.getProperties()),
                ProMatches.noCanalsBetweenTerritories(player, data));
```

I think there are some issues here:
1. DRY violation, the idea of how to calculate land-passable territories without canal is derived in two different places (happily the matching logic is in one spot, but still it's indirect to go from a matcher to a method call to actually getting the neighbors).
2. Makes for a difficult API to use. I don't think it's very fluent to look up a matcher and seems confusing to keep straight the difference between a route and territory condition. We also have a number of overloads which makes a person wonder how they should get neighbors if that is what they are trying to do.

What do you think we created an API that could hide the two predicates from callers?

The signature would be something like:
```
Set<Territory> getLandTraversableNeighbors(
      final Territory territory,
      final int distance);
```

Then in the javadoc we could define what is meant by land-traversable. Or we could call the method `getNonCanalLandTraversableNeighbors` if we feel the canal aspect is distinct enough to need to be in the method name. 

I would agree that having `ProMatches` in `GameMap` is bad, though I think we are already there indirectly by having entire methods (like this and similar) be in `GameMap` only to support AI. As noted earlier, the entire method likely ought to be picked up and moved to AI specific code.

Otherwise, WDYT?

Such a move should hopefully make the method overloads easier to use, reduce the API surface, reduce the number of parameters (which makes it more testable!), and just easier to understand how to use these methods.

Reply:
The idea sounds good but I'd rather not do it as part of this PR.  I think these types of changes need to be done after looking into all of the usages of `getNeighbor`.  I also think the canals check that I added probably needs to be used in more places than just these few places that I found.  It might turn out that it needs to be in all places that call `getNeighbor` with a territoryCondition.  But I'm not positive about it because of the size and scope of the AI code and my limited knowledge of its inner workings.

Reply:
I don't understand something. The checks  you added, the bi-predict, you have injected them from what is clearly AI code. What I'm suggesting to do is instead change the method the AI code calls and have the bi-predicates live in `GameMap`. That should not change the surface area of what you have updated and if there are other places that need to invoke the same logic it would also in turn invoke the new method on `GameMap` and would not duplicate the same Bi-predicates.

Reply:
I think it is premature to start trying to define the API in a different manner than how it is right now.  Yes, the existing API isn't that great but adding additional API points that are completely different will just make it even more confusing.  The API currently accepts conditions on the territories so adding a new parameter that is also a condition check isn't much different.

Reply:
I think we can do better than adding a parameter to an existing API. I disagree it is premature at all. We are setting a pattern. If another developer were to look at what is done here, "I need more functionality in this method - okay, add a parameter that only one part of the code uses and the rest will just pass a no-op value." Rinse-wash-repeat, and then eventually the concern of having too many params arise and the response is "the code was already like that, fixing all usages is out of scope, it is consistent to just add another parameter." Hence, I would say it's premature to add a parameter that make the method prematurely generic.

I don't see what it is a big change to encapsulate the match condition with the `getNeighbor` method. As-is, I think we violate DRY (which is about expressing an idea in one place, less about removing all duplicate patterns). Currently the neighbor logic is spread between match conditions only used for getNeighbor, the caller that invokes the method with the right parameter bindings, and then the logic in `GameMap`. Which is where the pattern is not a good one to follow, what if someone were to use this new public method API and pass in another match condition? It'd be an even worse situation, rather than the pattern to encapsulate the `getNeighbor` logic to one place.

Though again, I think we can do better than having a 4 param method and just add another method. We are debating so far between adding a 4 arg public method vs a 2 arg public method. The latter is readily tested, the former would likely involve heavy mocking and likely would not test very much. The heuristic of "can this be reasonably tested" I think fails when we consider almost any 4 arg method, this is an example I believe.

For more consideration, a predicate can be expressed as a bi-predicate, so we could keep the predicate logic all in the caller and add just one new API. EG:

```
public void predicateMethod(Predicate<Territory> territoryPredicate) {
   biPredicateMethod((start, end) -> territoryPredicate.apply(end));
}

public void biPredicateMethod(BiPredicate<Territory, Territory> routePredicate) {
   ///   
}
```

Reply:
I'll close this then since it doesn't appear to be the correct fix for the issue.

Reply:
If you want to work on it, go ahead.  I'm not going to do much else with it.

Reply:
Given the testing statement there is a fix here and that is very valuable. 

"Correct" design is in the eye of the beholder and depends on future events that nobody can know, we can only implement the best design that we see for current and expected future circumstances given a favorable ROI.

Are you going to pursue this fix any further @trevan ?

Reply:
@DanVanAtta , at this point, I'm not going to pursue any more here.  I'd love to get a more in-depth understanding of the AI code but the battle code is plenty big at this point in time.

Reply:
@trevan , just saw your last comment. That answers my question.

It appears you have lost patience for this update & process. I'm disappointed we have reached that point.

Let's be direct when we can to avoid any misconception. It's okay to say "no, those design changes are difficult and involved. I'm not sure the updates suggested can cleanly land, feel free to take a stab by adding a commit to this branch."

Already at a cursory glance, we are dealing with a mess:

![Screenshot from 2020-11-30 21-23-44](https://user-images.githubusercontent.com/12397753/100700673-dcd1bd80-3352-11eb-868e-568d91c1eac9.png)

Reply:
The AI code is an interesting beast!  I think that code belongs in a sub-project, and then ideally we could push a ton of supporting logic into that sub-project as well. I think that would clean up the core game-engine (game-core) considerably.

For example, WeakAI would be removed but it is injected as a dummy player into the battle logic. Keep an eye out for that, if you can sever the WeakAI from the OOL & battle logic & battle calculator, then we could drop that AI completely.